openfl-tutorials/interactive_api/PyTorch_Lightning_MNIST_GAN/workspace/PyTorch_Lightning_GAN.ipynb

```diff
@@ -0,0 +1,771 @@
+    "        return {\"val_loss\": val_real_loss}\n",
+    "\n",
+    "    def on_epoch_end(self):\n",
+    "        z = self.validation_z.type_as(self.generator.model[0].weight)\n",
+    "\n",
+    "        sample_imgs = self(z)\n",
+    "        grid = torchvision.utils.make_grid(sample_imgs)\n",
+    "        self.logger.experiment.add_image(\"generated_images\", grid, self.current_epoch)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5719fa50",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from pytorch_lightning.callbacks import Callback\n",
+    "\n",
+    "\n",
+    "class MetricsCallback(Callback):\n",
+    "    \"\"\"PyTorch Lightning metric callback.\"\"\"\n",
+    "\n",
+    "    def __init__(self):\n",
+    "        super().__init__()\n",
+    "        self.metrics = []\n",
+    "\n",
+    "    def on_epoch_end(self, trainer, pl_module):\n",
+    "        met = copy.deepcopy(trainer.callback_metrics)\n",
+    "        self.metrics.append(met)\n",
+    "\n",
+    "    def __call__(self):\n",
+    "        return self.get_callbacks()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "46692b08",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model = GAN(channels=1, width=28, height=28, train_disc_only=0, train_gen_only=0)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "greater-activation",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "optimizer = model.configure_optimizers()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "caroline-passion",
+   "metadata": {},
+   "source": [
+    "#### Register model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "handled-teens",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from copy import deepcopy\n",
+    "\n",
+    "# Need this plugin only if multiple optimizers are used. Not required for PyTorch Lightning with a single optimizer.\n",
+    "framework_adapter = (\n",
+    "    \"plugin_for_multiple_optimizers.FrameworkAdapterPluginforMultipleOpt\"\n",
+    ")\n",
+    "MI = ModelInterface(\n",
+    "    model=model, optimizer=optimizer, framework_plugin=framework_adapter\n",
+    ")\n",
+    "\n",
+    "initial_model = deepcopy(model)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "portuguese-groove",
+   "metadata": {},
+   "source": [
+    "### Define and register FL tasks"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "increasing-builder",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "TI = TaskInterface()\n",
+    "\n",
+    "import tqdm\n",
+    "\n",
+    "\n",
+    "# The Interactive API supports registering functions definied in main module or imported.\n",
+    "def function_defined_in_notebook(some_parameter):\n",
+    "    print(f\"Also I accept a parameter and it is {some_parameter}\")\n",
+    "\n",
```

Comment:
```suggestion

```

Reply:
@igor-davidyuk what needs to be changed here? Should I remove the comment?

Reply:
Let's remove the function itself as @igor-davidyuk  suggested